Helpers for gating stats

candy02058912 opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/dotcom-forge/issues/4592

> [!IMPORTANT]  
> Updated: The images below refer to the commit 64cf7e6e1fc14ac99c6ca8cf067b79cf9deba594, currently, we're shipping the hooks/helpers first.

## Proposed Changes

Hiding items still work:

https://github.com/Automattic/wp-calypso/assets/6586048/4a11c67c-0f2c-42e4-a055-7eb2beecb78e


![gating stats with hoc](https://github.com/Automattic/wp-calypso/assets/6586048/b1bf1fcd-951e-4889-919b-6497669e2427)

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* yarn test-client client/my-sites/stats/hooks/test/use-should-gate-stats.ts

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

candy02058912 committed:
Try gating paid stats with HOC

candy02058912 committed:
Remove testing to gate time picker code

lsl commented on client/my-sites/stats/site.jsx:
```diff
@@ -380,6 +380,7 @@ class StatsSite extends Component {
 					<div className={ moduleListClasses }>
 						<StatsModule
 							path="posts"
+							gridArea="posts"
```
Can we drop this in favor of path?

candy02058912 replied:
That's for making the grid layout to work properly, which sometimes is different from path

lsl reviewed (commented):
I like where this is going, can you figure out how to ship this early so we can iterate? 

Perhaps remove the usage and get a quick @Automattic/red review?

candy02058912 committed:
Remove withFeatureGating usage

candy02058912 commented:
> I like where this is going, can you figure out how to ship this early so we can iterate?
> 
> Perhaps remove the usage and get a quick @Automattic/red review?

We'll be shipping the helpers/hooks first.

candy02058912 committed:
Add utility function for stats feature gating

candy02058912 committed:
Add unit test

candy02058912 committed:
Remove redundant comments

lsl commented on client/my-sites/stats/hooks/use-feature-gating.ts:
```diff
@@ -0,0 +1,42 @@
+ * For example in mapStateToProps:
+ * const canAccessFeature = checkFeatureGating( state, siteId, 'statsSearchTerms' );
+ */
+export const checkFeatureGating = ( state: object, siteId: number | null, featureType: string ) => {
```
Are these statTypes a ts type anywhere we could use?

candy02058912 replied:
I dug around and found [this](https://github.com/Automattic/wp-calypso/blob/trunk/packages/wpcom.js/src/lib/runtime/site.get.js#L2). I'll add a file where we can make them into variables. Currently, `statType`s in calypso are hardcoded.

candy02058912 replied:
We could also do ⬆️ in a separate PR so that you can use the functions with your implementation. Feel free to merge if needed, and I'll iterate.

lsl replied:
Sure lets do that in a follow up. Don't worry too much about rushing this though, it's not super blocking and I think it's worth getting these things right the first time: https://github.com/Automattic/wp-calypso/pull/85041#issuecomment-1851059191

lsl reviewed (approved):
LGTM, @Automattic/red what's the right way to know if something needs testing against Jetpack?

obenland commented on client/my-sites/stats/hooks/use-feature-gating.ts:
```diff
@@ -0,0 +1,42 @@
+ * This hook is used to check if a stat feature is enabled for a site.
+ * Currently used for paid stats.
+ */
+export const useFeatureGating = ( featureType: string ) => {
+	const siteId = useSelector( getSelectedSiteId );
+	const canAccessFeature = useSelector( ( state ) =>
+		checkFeatureGating( state, siteId, featureType )
```
How would you feel about calling them `checkStatsFeatureGating` and use `useStatsFeatureGating` to make it clear on first glance that they are Stats-specific?

I worry that folks might want to use them instead of `siteHasFeature` for other things, maybe even change this hook in the future to broaden its use. It helps that it already lives in `my-sites/stats`, but I'd sleep better knowing that when folks read the imported name in code, that they don't get the wrong idea.

lsl replied:
More discussion below

grzegorz-cp commented:
Thanks for including the team in the review. It looks like a good start, but I've got a suggestion about the naming. `Feature gating` might be seen as a negative thing. What do you think about using something like a `premium feature`?

Since the hook might need specific checks that won't apply to Odyssey, can we reflect that in the name of the hook?

I think it would also be good to check for the site type and add an extra feature flag. Those checks would allow us to avoid affecting Odyssey and Jetpack on Calypso.

lsl commented:
> Thanks for including the team in the review. It looks like a good start, but I've got a suggestion about the naming. Feature gating might be seen as a negative thing. What do you think about using something like a premium feature?

Could you show how this might look in code? We currently have:

```
const { canAccessFeature } = useFeatureGating( featureType );
```

Would something like this work for you?

```
const { canAccessFeature } = useStatsFeatureAccess( featureType );
```

> Since the hook might need specific checks that won't apply to Odyssey, can we reflect that in the name of the hook?

I probably lack the needed background/context about Odyssey here. Could you elaborate on this? 

> I think it would also be good to check for the site type and add an extra feature flag. Those checks would allow us to avoid affecting Odyssey and Jetpack on Calypso.

I was under the impression @kangzj wanted to share Jetpack feature gating with dotcom but we can add site type checking. I don't think we need to pass that in though, I suspect it would be an internal-to-the-hook check.

> Those checks would allow us to avoid affecting Odyssey and Jetpack on Calypso.

How should we test this? I'm not sure I understand the odyssey/Jetpack setup, I thought these were more or less the same thing.

candy02058912 commented:
> I think it would also be good to check for the site type and add an extra feature flag. Those checks would allow us to avoid affecting Odyssey and Jetpack on Calypso.

Quick question regarding this, atomic sites also use Jetpack on Calypso and not Odyssey right?
The site type can be an internal checked inside the `checkFeatureGating` (official name TBD)

> How should we test this? I'm not sure I understand the odyssey/Jetpack setup, I thought these were more or less the same thing.

+1 to this

candy02058912 committed:
Update tests and function names

candy02058912 commented:
I tried updating the function names to accommodate for the comments, please give it another review, thanks!

lsl commented:
> and reflecting it's a wpcom thing

I don't agree with this but I want to move ahead with what we have, lets go with this for now and change the name if Jetpack wants to use the same feature gating at a later date.

kangzj commented:
Thanks for looping me in.

> I don't agree with this but I want to move ahead with what we have, lets go with this for now and change the name if Jetpack wants to use the same feature gating at a later date.

I kinda think it's necessary, because Jetpack is going to build feature gating as well, so if it's a general name, then it'll be very much confusing for devs from both teams.

>     How should we test this? I'm not sure I understand the odyssey/Jetpack setup, I thought these were more or less the same thing.

Here's a FG for it: PCYsg-Pp7-p2

@candy02058912 @lsl

kangzj commented on client/my-sites/stats/hooks/use-wpcom-paid-stats.ts:
```diff
@@ -0,0 +1,59 @@
+
+	// check feature flags
+	if ( ! isPaidStatsEnabled || isOdysseyStats ) {
+		return false;
```
I don't know. I feel like if it's the sites that the strategy doesn't want to gate, then should we return `true` instead of `false`? Or a `null` to identify that it's not applicable?

candy02058912 replied:
Should we still be checking for OdysseyStats? I'm a bit confused by this comment.

lsl replied:
Needs to return true here to disable the paywall gating when the config flag is not set or it's an odyssey context. Currently this is paywalling my jetpack test site with https://github.com/Automattic/wp-calypso/pull/85126/ applied.

This says you don't have access to a feature (false value) if you're using Jetpack or paid stats is disabled.

Maybe we should name it something like "shouldPaywall" or "shouldGateFeature"?

candy02058912 replied:
Decided to change the names to not make things confusing, WIP

candy02058912 commented on client/my-sites/stats/highlights-section/index.tsx:
```diff
@@ -15,13 +15,7 @@ import { requestHighlights } from 'calypso/state/stats/highlights/actions';
-	siteId: number;
-	currentPeriod: string;
-} ) {
+function HighlightsSection( { siteId, currentPeriod }: { siteId: number; currentPeriod: string } ) {
```
TODO: remove changes for this file in order to not cause confusion

lsl commented:
> I kinda think it's necessary, because Jetpack is going to build feature gating as well, so if it's a general name, then it'll be very much confusing for devs from both teams.

We should be building the same feature gating :smiley:

kangzj commented:
> We should be building the same feature gating

I'm very keen... but with different products, and different pricing strategies, I don't know how this would be possible 🤔

lsl commented:
It should be the same stats/features that get gated behind the addon/feature purchase right?

Pricing, packaging, tiered or not seems separate from what stats/features get gated.

We can also change this code to accommodate more complicated gating if Jetpack's packaging requires it. At most it'll probably just be some kind of product feature that changes the list of features being gated.

candy02058912 committed:
revert HighlightsSection changes

candy02058912 committed:
Update naming and tests to make the functions more clear

candy02058912 committed:
Update function comment for useShouldGateStats

candy02058912 commented:
I decided to merge this PR first, but this PR **can still be reviewed post-merge.**

Doing it to help move https://github.com/Automattic/wp-calypso/pull/85126 forward and see how everything is like after being integrated.

More context: 
I got confused with the naming myself, so I've updated the naming and adjusted the comments. (please check the files) Also tried to make the unit tests more explanatory and useful.

Couldn't come up with a better name that fulfilled all of the comments in this PR and not make it overly long.
